chunk.go

```diff
@@ -0,0 +1,159 @@
+package iavl
+
+import (
+	"sort"
+)
+
+// Chunk is a list of ordered nodes
+// It can be sorted, merged, exported from a tree and
+// used to generate a new tree
+type Chunk []OrderedNodeData
+
+// OrderedNodeData is the data to recreate a leaf node,
+// along with a SortOrder to define a BFS insertion order
+type OrderedNodeData struct {
+	SortOrder uint64
+	NodeData
+}
+
+// NewOrderedNode creates the data from a leaf node
+func NewOrderedNode(leaf *Node, prefix uint64) OrderedNodeData {
+	return OrderedNodeData{
+		SortOrder: prefix,
+		NodeData: NodeData{
+			Key:   leaf.key,
+			Value: leaf.value,
+		},
+	}
+}
+
+// GetChunkHashes returns all the "checksum" hashes for
+// the chunks that will be sent
+func GetChunkHashes(tree *Tree, depth uint) ([][]byte, uint) {
+	// Account for unbalanced trees by capping the depth.
+	maxDepth := uint(tree.root.height / 2)
+	if depth > maxDepth {
+		depth = maxDepth
+	}
+
+	nodes := getNodes(tree, depth)
+	res := make([][]byte, len(nodes))
+	for i, n := range nodes {
+		res[i] = n.hash
+	}
+	return res, depth
+}
+
+// getNodes returns an array of nodes at the given depth
```

Comment:
doesnt this actually fetch all nodes "up to" the given depth (not "at" the given depth) ?

Reply:
Look at traverseDepth, it only returns when depth==0
It just calls left and right at depth>0, without adding any other info